From 29d7d7411d74ddd136f880af4a30d508bc04da60 Mon Sep 17 00:00:00 2001
From: Naveen S <quic_naves@quicinc.com>
Date: Wed, 4 Sep 2024 23:03:44 +0530
Subject: [PATCH] sigma_dut: Restructure ap_set_refeature

In channel switch, the swapped link id will
be used in hostapd_cli earlier, now removing
the swapping part as it has been handled correctly
now.

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Signed-off-by: Naveen S <quic_naves@quicinc.com>
---
 ap.c | 12 +++---------
 1 file changed, 3 insertions(+), 9 deletions(-)

--- a/ap.c
+++ b/ap.c
@@ -17822,7 +17822,6 @@ static int mac80211_chnum_band(struct si
 	char link_param[10];
 	int res;
 	bool link_set = false;
-	int link = 0;
 
 	/* Extract the channel info */
 	token = strdup(val);
@@ -17838,12 +17837,7 @@ static int mac80211_chnum_band(struct si
 		chwidth = atoi(result);
 
 	if (dut->num_links > 1) {
-		if (link_id == 0)
-			link = 1;
-		else
-			link = 0;
-
-		center_freq_idx = get_oper_centr_freq_seq_idx(dut, chwidth, channel, link);
+		center_freq_idx = get_oper_centr_freq_seq_idx(dut, chwidth, channel, link_id);
 	} else {
 		center_freq_idx = get_oper_centr_freq_seq_idx(dut, chwidth, channel, 0);
 	}
@@ -17858,7 +17852,7 @@ static int mac80211_chnum_band(struct si
 	else
 		sec_ch_offset = 1;
 
-	if (check_band(dut, BAND_2G, link)) {
+	if (check_band(dut, BAND_2G, link_id)) {
 		if (chwidth == 40) {
 			if (channel < 7)
 				sec_ch_offset = 1;
@@ -17867,7 +17861,7 @@ static int mac80211_chnum_band(struct si
 		}
 		center_freq = get_2g_channel_freq(center_freq_idx);
 		channel_freq = get_2g_channel_freq(channel);
-	} else if (check_band(dut, BAND_6G, link)) {
+	} else if (check_band(dut, BAND_6G, link_id)) {
 		if (chwidth != 20) {
 			if (((channel - 1) / 4) % 2)
 				sec_ch_offset = -1;
@@ -17945,7 +17939,7 @@ static void mac80211_prepare_nr_buf(stru
 		sizeof(nr_buf) - strlen(nr_buf));
 }
 
-static int mac80211_update_mbo_pref_ap_list(struct sigma_dut *dut)
+static int mac80211_update_mbo_pref_ap_list(struct sigma_dut *dut, int link_id)
 {
 	int i;
 	int least_pref = 1 << 8;
@@ -17991,9 +17985,9 @@ static int mac80211_update_mbo_pref_ap_l
 
 	/* Now add the self AP Address */
 	if (dut->mbo_self_ap_tuple.ap_ne_class == -1) {
-		if (check_band(dut, BAND_6G, 0))
+		if (check_band(dut, BAND_6G, link_id))
 			ap_ne_class = 131;
-		else if (dut->ap_channel <= 11)
+		else if (dut->mld[link_id].ap_channel <= 11)
 			ap_ne_class = 81;
 		else
 			ap_ne_class = 115;
@@ -18003,7 +17997,7 @@ static int mac80211_update_mbo_pref_ap_l
 	}
 
 	if (dut->mbo_self_ap_tuple.ap_ne_op_ch == -1) {
-		ap_ne_op_ch = dut->ap_channel;
+		ap_ne_op_ch = dut->mld[link_id].ap_channel;
 		dut->mbo_self_ap_tuple.ap_ne_op_ch = ap_ne_op_ch;
 	} else {
 		ap_ne_op_ch = dut->mbo_self_ap_tuple.ap_ne_op_ch;
@@ -18035,7 +18029,7 @@ static int mac80211_update_mbo_pref_ap_l
 }
 
 static int mac80211_set_nebor(struct sigma_dut *dut, const char *ifname,
-			      struct sigma_cmd *cmd)
+			      struct sigma_cmd *cmd, int link_id)
 {
 	const char *val;
 	/*
@@ -18191,7 +18185,7 @@ static int mac80211_set_nebor(struct sig
 				dut->btm[neighbor_cnt].nebor_op_class[0],
 				dut->btm[neighbor_cnt].nebor_op_ch[0],
 				dut->btm[neighbor_cnt].ap_ne_pref);
-		if (mac80211_update_mbo_pref_ap_list(dut))
+		if (mac80211_update_mbo_pref_ap_list(dut, link_id))
 			return -1;
 	}
 
@@ -18304,10 +18298,11 @@ static void mac80211_set_trigger_type_1(
 	mac80211_set_mu_brp_commands(dut, ifname);
 }
 
-static void mac80211_set_trigger_type_2(struct sigma_dut *dut, const char *ifname)
+static void mac80211_set_trigger_type_2(struct sigma_dut *dut, const char *ifname,
+					int link_id)
 {
 	/* TriggerType "2" for MU BAR */
-	if (dut->ap_channel >= 36) {
+	if (dut->mld[link_id].ap_channel >= 36) {
 		/* RU allocation RU 242 - DL OFDMA data */
 		run_system_wrapper(dut,
 				   "ath11k-fwtest -i %s -m 0x4b -v 0 1 0 3 1 3 2 3 3 3",
@@ -18381,7 +18376,7 @@ static int mac80211_set_trigger_type(str
 		mac80211_set_trigger_type_1(dut, ifname);
 		break;
 	case 2:
-		mac80211_set_trigger_type_2(dut, ifname);
+		mac80211_set_trigger_type_2(dut, ifname, link_id);
 		break;
 	case 3:
 		mac80211_set_trigger_type_3(dut, ifname);
@@ -18398,7 +18393,8 @@ static int mac80211_set_trigger_type(str
 
 static enum sigma_cmd_result
 mac80211_tx_bandwidth(struct sigma_dut *dut, struct sigma_conn *conn,
-		      const char *ifname, const char *val, const char *type)
+		      const char *ifname, const char *val,
+		      const char *type, int link_id)
 {
 	int width, center_freq_idx, center_freq, channel_freq, res;
 	char *mode, buf[256];
@@ -18425,12 +18421,12 @@ mac80211_tx_bandwidth(struct sigma_dut *
 
 	width = atoi(val);
 	center_freq_idx = get_oper_centr_freq_seq_idx(dut, width,
-						      dut->ap_channel, 0);
+						      dut->mld[link_id].ap_channel, 0);
 	if (center_freq_idx < 0)
 		return ERROR_SEND_STATUS;
 
 	center_freq = get_5g_channel_freq(center_freq_idx);
-	channel_freq = get_5g_channel_freq(dut->ap_channel);
+	channel_freq = get_5g_channel_freq(dut->mld[link_id].ap_channel);
 
 	res = snprintf(buf, sizeof(buf),
 		       "CHAN_SWITCH 10 %d sec_channel_offset=1 center_freq1=%d bandwidth=%d blocktx %s",
@@ -18529,7 +18525,7 @@ static enum sigma_cmd_result mac80211_he
 	int ret = -1;
 	int ltf = -1, gi = -1;
 
-	if (dut->ap_is_dual && dut->ap_mode == AP_11be) {
+	if (dut->ap_is_dual && dut->mld[link_id].ap_mode == AP_11be) {
 		if (link_id == dut->band_to_link[0])
 			mode = "2.4";
 		else if (link_id == dut->band_to_link[1])
@@ -18743,7 +18739,7 @@ static enum sigma_cmd_result mac80211_eh
 			fwtest_linkid_cmd_wrapper(dut, "-m 0x4b -v 0 40 0 996", ifname, linkid);
 			fwtest_linkid_cmd_wrapper(dut, "-m 0x4b -v 0 41 0 996", ifname, linkid);
 		} else if (val_bw >= AP_160) {
-			if (dut->ap_chwidth == AP_320) {
+			if (dut->mld[linkid].ap_chwidth == AP_320) {
 				fwtest_linkid_cmd_wrapper(dut,
 							  "-m 0x4b -v 0 40 0 996 1 996 2 996 3 996",
 							  ifname, linkid);
@@ -18763,7 +18759,7 @@ static enum sigma_cmd_result mac80211_eh
 		}
 	} else if (value == EHT_RU_ALLOC_2) {
 		if (val_bw == AP_160) {
-			if (dut->ap_chwidth == AP_320) {
+			if (dut->mld[linkid].ap_chwidth == AP_320) {
 				fwtest_linkid_cmd_wrapper(dut, "-m 0x4b -v 0 40 0 1992 1 1992",
 							  ifname, linkid);
 				fwtest_linkid_cmd_wrapper(dut, "-m 0x4b -v 0 41 0 1992 1 1992",
@@ -19063,7 +19059,7 @@ static enum sigma_cmd_result mac80211_ap
 	if (val && mac80211_chnum_band(dut, ifname, val, link_id) < 0)
 		return -1;
 
-	mac80211_set_nebor(dut, ifname, cmd);
+	mac80211_set_nebor(dut, ifname, cmd, link_id);
 
 	val = get_param(cmd, "Assoc_Disallow");
 	if (val)
@@ -19109,7 +19105,7 @@ static enum sigma_cmd_result mac80211_ap
 	val = get_param(cmd, "txBandwidth");
 	if (val) {
 		res = mac80211_tx_bandwidth(dut, conn, ifname, val,
-					       get_param(cmd, "type"));
+					    get_param(cmd, "type"), link_id);
 		if (res != SUCCESS_SEND_STATUS)
 			return res;
 	}
@@ -19201,7 +19197,7 @@ static enum sigma_cmd_result mac80211_ap
 	val = get_param(cmd, "CodingType");
 	if (val) {
 		if (strcasecmp(val, "BCCCoding") == 0 && dut->program == PROGRAM_EHT) {
-			if (dut->ap_mode == AP_11be) {
+			if (dut->mld[link_id].ap_mode == AP_11be) {
 				if (dut->ap_fixed_rate) {
 					dut->ap_bcc = VALUE_ENABLED;
 					dut->ap_ldpc = VALUE_DISABLED;
@@ -19214,7 +19210,7 @@ static enum sigma_cmd_result mac80211_ap
 							  "-t 1 -m 0x00 -v 0  0x8A 0",
 							  ifname, link_id);
 		} else if (strcasecmp(val, "LDPCCoding") == 0 && dut->program == PROGRAM_EHT) {
-			if (dut->ap_mode == AP_11be) {
+			if (dut->mld[link_id].ap_mode == AP_11be) {
 				if (dut->ap_fixed_rate)
 					dut->ap_ldpc = VALUE_ENABLED;
 				else
@@ -19273,7 +19269,7 @@ static enum sigma_cmd_result mac80211_ap
 
 			punctchannel_val = strtok_r(punct_channel, " ", &saveptr);
 
-			if (dut->ap_link_chwidth[link_id] == AP_80) {
+			if (dut->mld[link_id].ap_chwidth == AP_80) {
 				int punct_80[] = {33, 37, 41, 45};
 
 				chwidth = 80;
@@ -19285,7 +19281,7 @@ static enum sigma_cmd_result mac80211_ap
 					}
 					punctchannel_val = strtok_r(NULL, " ", &saveptr);
 				}
-			} else if (dut->ap_link_chwidth[link_id] == AP_160) {
+			} else if (dut->mld[link_id].ap_chwidth == AP_160) {
 				int punct_160[] = {33, 37, 41, 45, 49, 53, 57, 61};
 
 				chwidth = 160;
@@ -19332,8 +19328,8 @@ static enum sigma_cmd_result mac80211_ap
 						dut->ap_punct_bitmap);
 			}
 
-			chan = channel_to_freq(dut, dut->ap_channel);
-			center_freq_idx = get_oper_center_freq_6g(chwidth, dut->ap_channel);
+			chan = channel_to_freq(dut, dut->mld[link_id].ap_channel);
+			center_freq_idx = get_oper_center_freq_6g(chwidth, dut->mld[link_id].ap_channel);
 			center_freq = channel_to_freq(dut, center_freq_idx);
 
 			if (dut->cfreq == 0)
@@ -19343,11 +19339,11 @@ static enum sigma_cmd_result mac80211_ap
 			else
 				sec_ch_offset = -1;
 
-			center_freq_idx = get_oper_center_freq_6g(chwidth, dut->ap_channel);
+			center_freq_idx = get_oper_center_freq_6g(chwidth, dut->mld[link_id].ap_channel);
 			center_freq = channel_to_freq(dut, center_freq_idx);
 
 			res = snprintf(buf, sizeof(buf),
-					" -i %s %s chan_switch 10 %d sec_channel_offset=%d center_freq1=%d bandwidth=%d punct_bitmap=%d blocktx %s",
+					" -i %s %s chan_switch 10 %d sec_channel_offset=%d center_freq1=%d bandwidth=%d punct_bitmap=%d blocktx %s",
 				ifname, link_set ? link_param : "", chan,
 				sec_ch_offset, center_freq, chwidth, dut->ap_punct_bitmap, "eht");
 
@@ -19473,15 +19469,15 @@ static enum sigma_cmd_result mac80211_ap
 
 	val = get_param(cmd, "MRUAllocTones");
 	if (val) {
-		if (dut->ap_link_chwidth[link_id] == AP_20)
+		if (dut->mld[link_id].ap_chwidth == AP_20)
 			val_bw = 0;
-		else if (dut->ap_link_chwidth[link_id] == AP_40)
+		else if (dut->mld[link_id].ap_chwidth == AP_40)
 			val_bw = 1;
-		else if (dut->ap_link_chwidth[link_id] == AP_80)
+		else if (dut->mld[link_id].ap_chwidth == AP_80)
 			val_bw = 2;
-		else if (dut->ap_link_chwidth[link_id] == AP_160)
+		else if (dut->mld[link_id].ap_chwidth == AP_160)
 			val_bw = 3;
-		else if (dut->ap_link_chwidth[link_id] == AP_320)
+		else if (dut->mld[link_id].ap_chwidth == AP_320)
 			val_bw = 4;
 		else
 			val_bw = -1;
@@ -19518,15 +19514,15 @@ static enum sigma_cmd_result mac80211_ap
 		if (value_bw) {
 			val_bw = atoi(value_bw);
 		} else {
-			if (dut->ap_link_chwidth[link_id] == AP_20)
+			if (dut->mld[link_id].ap_chwidth == AP_20)
 				val_bw = 0;
-			else if (dut->ap_link_chwidth[link_id] == AP_40)
+			else if (dut->mld[link_id].ap_chwidth == AP_40)
 				val_bw = 1;
-			else if (dut->ap_link_chwidth[link_id] == AP_80)
+			else if (dut->mld[link_id].ap_chwidth == AP_80)
 				val_bw = 2;
-			else if (dut->ap_link_chwidth[link_id] == AP_160)
+			else if (dut->mld[link_id].ap_chwidth == AP_160)
 				val_bw = 3;
-			else if (dut->ap_link_chwidth[link_id] == AP_320)
+			else if (dut->mld[link_id].ap_chwidth == AP_320)
 				val_bw = 4;
 			else
 				val_bw = -1;
